File: study/src/gawonkim/Baekjoon/boj_5639.c
#include <stdio.h>
#include <stdlib.h>
typedef struct TreeNode
{
    struct TreeNode *left;
    struct TreeNode *right;
    int key;
} TreeNode;

TreeNode *createNode(int data)
{
    TreeNode *newNode = (TreeNode *)malloc(sizeof(TreeNode));
    newNode->key = data;
    newNode->left = NULL;
    newNode->right = NULL;
    return newNode;
}

TreeNode *insertNode(TreeNode *node, int data)
{
    if (node == NULL)
        return createNode(data);
    else if (node->key > data)
        node->left = insertNode(node->left, data);
    else if (node->key < data)
        node->right = insertNode(node->right, data);
    return node;
}

void postorder(TreeNode *root)
{
    if (root)
    {
        postorder(root->left);
        postorder(root->right);
        printf("%d\n", root->key);
    }
}

int main(void)
{
    int input;
    TreeNode *root = NULL;
    while (scanf("%d", &input) != EOF)
        root = insertNode(root, input);

    postorder(root);

    return 0;
}

File: study/src/gawonkim/Baekjoon/boj_5639.py
import sys

class TreeNode:
    def __init__(self, key):
        self.key = key
        self.left = None
        self.right = None

def insert_node(node, key):
    if node is None:
        return TreeNode(key)
    if key < node.key:
        node.left = insert_node(node.left, key)
    elif key > node.key:
        node.right = insert_node(node.right, key)
    return node

def postorder(root):
    if root:
        postorder(root.left)
        postorder(root.right)
        print(root.key)

if __name__ == "__main__":
    root = None
    while True:
        input_val = sys.stdin.readline().strip()
        if not input_val:
            break
        try:
            input_val = int(input_val)
            root = insert_node(root, input_val)
        except EOFError:
            break
    
    postorder(root)